### PR TITLE
Feature/dropout ppca

### PR DESCRIPTION
**Pull request by austinTalbot7241993**
This is probabilistic pca but with a supervised variational objective added, can be classification or regression. Currently built by extending PPCA object previously added.

### CODE REVIEW

**Review by poneill — CHANGES_REQUESTED** [review]
Nice work!  Let's definitely fix the deprecation warning and move away from full model training runs in unit tests, though.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+SPPCADropout(SPCA)
+    This is SPCA with SGD but supervises a single factor. 
+
+TODO
+FactorAnalysisDropout(FactorAnalysis)
+    This is factor analysis but supervises a single factor.
+
+Methods
+-------
+_get_projection_matrix(W_,sigma_,n_components)
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:22]: OK yeah that makes sense. Will need to eventually change anyways to handle beyond PPCA.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+from bystro.supervised_ppca._misc_np import softplus_inverse_np
+from bystro.supervised_ppca.gf_generative_pt import PPCA
+
+
+def _get_projection_matrix(W_, sigma_):
+    """
+    This is currently just implemented for PPCA due to nicer formula. Will
+    modify for broader application later.
+
+    Computes the parameters for p(S|X)
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:43]: OK, we need to get on publishing the paper lol. I'll put a link to give incentive.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    Returns
+    -------
+    Proj_X : pt.tensor(n_components,p)
+        Beta such that Proj_XX = E[S|X]
+
+    Cov : pt.tensor(n_components,n_components)
+        Var(S|X)
+    """
+    n_components = int(W_.shape[0])
+    eye = torch.tensor(np.eye(n_components))  # .astype(np.float32))
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:62]: Surprised that ruff missed that.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+                self.delta
+                * torch.matmul(z_samples[:, : self.n_supervised], one_s)
+                + B_
+            )
+            if task == "regression":
+                loss_y = supervision_loss(y_hat, y_batch)
+            else:
+                loss_y = supervision_loss(sigm(y_hat), y_batch)
+
+            WTW = torch.matmul(W_, torch.transpose(W_, 0, 1))
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:215]: Ah right, so it depends on whether W\in R^{Lxp} or R^{pxL}, and in the code implementation I match sklearn which is the former.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        super().__init__(
+            n_components=n_components,
+            prior_options=prior_options,
+            training_options=training_options,
+        )
+        self._initialize_save_losses()
+        self.losses_supervision = np.empty(
+            self.training_options["n_iterations"]
+        )
+
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:120]: As none?

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+        sigmal_ = torch.tensor(sinv, requires_grad=True)
+        B_ = torch.tensor(0.0, requires_grad=True)
+        return W_, sigmal_, B_
+
+    def _test_inputs(self, X, y):
+        """
+        Just tests to make sure data is numpy array and dimensions match
+        """
+        if not isinstance(X, np.ndarray):
+            raise ValueError("Data is numpy array")
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:293]: What?

[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:293]: Oh the comment sure

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        ----------
+        trainable_variables : list
+            List of tensorflow variables saved
+
+        Sets
+        ----
+        W_ : np.array-like,(n_components,p)
+            The loadings
+
+        sigmas_ : np.array-like,(n_components,p)
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:246]: No or at least they aren't the same. Sigmal_ is the latent unconstrained variable we optimize with respect to. Sigma2_ is the isotropic variance of PPCA while sigmas_ is the not isotropic diagonal in factor analysis.

File: python/python/bystro/supervised_ppca/tests/test_dropout_pt.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+    logits = phi * S[:, 0] * 5
+    probs = np.exp(logits) / (1 + np.exp(logits))
+    y = binom.rvs(1, probs, random_state=42)
+    return X, y, X_hat, S, W, logits
+
+
+def test_ppca():
+    X, y, X_hat, S, W, logits = PPCA_generate_data(L=3, p=400)
+    training_options = {"n_iterations": 3000}
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/tests/test_dropout_pt.py:29]: (2)

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:215]: I agree with defining `W` that way (i.e. having dimensions of (observations x features)) but I'm still confused about why we're naming this variable `WTW` (which, to my mind, strongly suggests `W.T @ W`) when we're computing `W @ W.T`?

Does my question make sense?

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:120]: Yes.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:246]: Maybe I can try to rephrase my question:

The docstring for this method says that it sets three variables, `W_`, `sigmas_` and `B_`.  But the method actually sets `W_`, `sigma2_` and `B_`.  

If there's a discrepancy between what the docstring says the method does and what the method does, we should resolve this discrepancy so that future developers aren't confused.

On the other hand, if the docstring and method are in fact correct as written (i.e. the docstring should say we set `sigmas_` and the method should set `sigma2_`), that's probably worth a comment explaining the distinction.

AIUI this class--because it implements PPCA--shouldn't have even have a `sigmas_` attribute, right?

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+    W_ : pt.Tensor(n_components,p)
+        The loadings
+
+    sigma_ : pt.Flaot
+        Isotropic noise
+
+    Returns
+    -------
+    Proj_X : pt.tensor(n_components,p)
+        Beta such that Proj_XX = E[S|X]
```
[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:56]: As written, it looks like a single variable `Proj_XX`.

In general, questions that come up during code review are likely to come up again when other engineers read this code-- could we please default to answering questions *through the code* so that future engineers get the benefit of these clarifications as well?  It's very unlikely that anyone will look back at this code review when reading this code, but if we just amend `Proj_XX` to `Proj_X @ X` (or similar) they won't have to :)

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+                + B_
+            )
+            if task == "regression":
+                loss_y = supervision_loss(y_hat, y_batch)
+            else:
+                loss_y = supervision_loss(sigm(y_hat), y_batch)
+
+            WTW = torch.matmul(W_, torch.transpose(W_, 0, 1))
+            off_diag = WTW - torch.diag(torch.diag(WTW))
+            torch.norm(off_diag)
```
Comment (poneill) [inline python/python/bystro/supervised_ppca/gf_dropout_pt.py:217]: what's going on here? This line appears to be a no-op, and torch.norm also seems to be deprecated: https://pytorch.org/docs/stable/generated/torch.norm.html

Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:217]: Apparently "ruff --fix" isn't particularly reliable. I'll fix it and push again with matix_norm. But it's forcing independence between the different components, happens naturally in PPCA but not inherent with supervision loss.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:246]: Yeah that's correct. Are we planning on making docstrings standard in all modules in Bystro?

File: python/python/bystro/supervised_ppca/tests/test_dropout_pt.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/tests/test_dropout_pt.py:29]: Sorry, my bad, on further reflection I don't think (2) is going to work, especially if we continue in this vein, because we're still going to have to run everything on CI.

Would you agree that it's a reasonable goal to have the test suite finish in a single-digit number of minutes?  Asking because we're almost hitting that ceiling now.

If we're going to do train for more than a single-digit number of iterations we should do so with a tiny model on a tiny dataset.  Would we lose substantive by reducing the number of latent dimensions to 1 or 2 here?

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (akotlar) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:246]: > Yeah that's correct. Are we planning on making docstrings standard in all modules in Bystro?

Yeah, as you and I chatted offline, for public methods, I think we should, since we'll be auto-generating documentation off of it.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:215]: Because the paper I will write conflicts with sklearn.

File: python/python/bystro/supervised_ppca/tests/test_dropout_pt.py
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/tests/test_dropout_pt.py:1]: BTW the convergence error is on not our method for a unit test so it doesn't matter.

[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/tests/test_dropout_pt.py:1]: I think it does matter: having unit tests that pass cleanly (i.e. without a bunch of warnings) is important for a few reasons:

1.  It's not obvious that the warning is unrelated to the behavior of the test.  If the test later fails and the engineer responsible sees that the failing test is also generating a warning about convergence error, that would be a natural place to start looking.
2. Having a lot of uninformative warnings in unit tests train us to stop paying attention to the output of unit tests, which means it will be easier to miss warnings that do matter.
3.  If the warning happens in testing, it's likely to happen in production as well, and whoever gets that warning will probably have even less context to deal with it correctly.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:221]: Sorry, I don't think I'm following you here.  It sounds like you're saying you want to deliberately name a variable contrary to its definition, in a way that looks exactly like an innocent typo / thinko, because that would agree with envisioned notation in an as-of-yet unwritten paper?

I feel like I must be misunderstanding you here-- do you understand my concern about writing `WTW = W @ W.T` and could you help me to understand why it's mistaken, if it is?

[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:221]: It reflects the fact that in a program it's easier to have W \in R^{Lxp} but in mathematics have W \in R^{pxL} and I match math papers not code.

File: python/python/bystro/supervised_ppca/tests/test_dropout_pt.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/tests/test_dropout_pt.py]: The program part of unit tests takes 2.5 minutes, most of it I have no control over.

File: python/python/bystro/supervised_ppca/tests/test_dropout_pt.py
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/tests/test_dropout_pt.py:1]: The warning says it's an sklearn warning.

[the remark this answers is not in the document]
Comment (akotlar) [reply python/python/bystro/supervised_ppca/tests/test_dropout_pt.py:1]: I no longer see any warnings

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (akotlar) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:230]: With the comment at the top of this file explaining the convention, I’m satisfied.

**Review by akotlar — APPROVED** [review]
Looks good. No blocking comments from me.

File: python/python/bystro/supervised_ppca/tests/test_dropout_pt.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+    logits = phi * S[:, 0] * 5
+    probs = np.exp(logits) / (1 + np.exp(logits))
+    y = binom.rvs(1, probs, random_state=42)
+    return X, y, X_hat, S, W, logits
+
+
+def test_ppca():
+    X, y, X_hat, S, W, logits = PPCA_generate_data(L=3, p=200)
+    training_options = {"n_iterations": 1000}
```
Comment (poneill) [inline python/python/bystro/supervised_ppca/tests/test_dropout_pt.py:29]: Please be careful with these: they can lead to qualitative slow-downs in development velocity for everyone else.  If we're unit testing more than a few iterations of a pytorch model we're performing the wrong test.  

Is 1000 iterations demonstrably necessary here or would a smaller number suffice?

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+        Parameters
+        ----------
+        X : np.array-like,(n_samples,n_covariates)
+            The data
+
+        y : np.array-like,(n_samples,n_prediction)
+            Covariates we wish to predict. For now lazy and assuming
+            logistic regression.
+
+        task : string,default='classification'
```
Comment (poneill) [inline python/python/bystro/supervised_ppca/gf_dropout_pt.py:147]: it would be helpful to explicitly list the possible alternatives here (is the string "multinomial regression" a valid input?) and use a `match` statement to explicitly do the right thing for each possible value of `task` every time we reference it.  This line seems to disagree with the block ending on L176, which doesn't make any mention of multinomial regression.

You can have mypy spot these errors for you by defining the valid tasks as follows:

`Task = Literal["classification", "prediction", "multinomial_regression"]`

then annotating the input as type `Task` and ending each if-elif-block with an else with a call to assert_never():

```python
def assert_never(value: NoReturn) -> NoReturn:
    assert False, f"This code should never be reached, got: {value}"
```

Or you could do it by hand but the trade-off is that you have to worry about keeping everything in sync.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+    eye = torch.tensor(np.eye(n_components))
+    M = torch.matmul(W_, torch.transpose(W_, 0, 1)) + sigma_ * eye
+    Proj_X = torch.linalg.solve(M, W_)
+    Cov = torch.linalg.inv(M) * sigma_
+    return Proj_X, Cov
+
+
+class PPCADropout(PPCA):
+    """
+    This implements supervised PPCA according to the paper draft that I'm
```
Comment (poneill) [inline python/python/bystro/supervised_ppca/gf_dropout_pt.py:80]: This docstring is part of the public API for the class and shouldn't contain the first-person pronoun because its referent is ambiguous.  Instead: "according to Talbot et al. (in prep, 2023) 'Prospective Title Goes Here'" so that if this docstring isn't updated in a year, the reader has some hope of finding it.

**Review by poneill — APPROVED** [review]
Unblocking this PR but would recommend clearly defining the set of possible `tasks` in the `fit` method of `PPCADropout` and ensuring that they're all accounted for each time we branch on the value of `task`.  This affects the public API of PPCADropout and so will be harder to revise later than a private implementation detail would be.  One way of getting the tooling to check this for you is suggested below.

Please also set `n_iterations` to the lowest value that gets the test passing and revise the test criterion if necessary.

File: python/python/bystro/supervised_ppca/tests/test_dropout_pt.py (hunk 11 above)
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/tests/test_dropout_pt.py:29]: That's a discussion, do we want unit tests to evaluate performance in realistic scenarios, in which case probably yes.

File: python/python/bystro/supervised_ppca/gf_dropout_pt.py (hunk 12 above)
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/gf_dropout_pt.py:147]: In the next PR will add support for multinomial regression, but would be nice to leave in separate pr.
